src/go/k8s/internal/controller/redpanda/user_controller.go

```diff
@@ -0,0 +1,229 @@
+		if err := usersClient.Create(ctx, user); err != nil {
+			return handleErrors(err)
+		}
+		hasManagedUser = true
+	}
+
+	if hasUser && !shouldManageUser {
+		if err := usersClient.Delete(ctx, user); err != nil {
+			return handleErrors(err)
+		}
+		hasManagedUser = false
+	}
+
+	if shouldManageACLs {
+		if err := syncer.Sync(ctx, user); err != nil {
+			return handleErrors(err)
+		}
+		hasManagedACLs = true
+	}
+
+	if !shouldManageACLs && hasManagedACLs {
+		if err := syncer.DeleteAll(ctx, user); err != nil {
+			return handleErrors(err)
+		}
+		hasManagedACLs = false
+	}
+
+	return redpandav1alpha2.UserSyncedCondition(user.Name), hasManagedUser, hasManagedACLs, nil
+}
+
+func (r *UserReconciler) deleteUser(ctx context.Context, user *redpandav1alpha2.User) error {
+	hasManagedACLs, hasManagedUser := user.HasManagedACLs(), user.HasManagedUser()
+
+	ignoreAllConnectionErrors := func(err error) error {
+		// If we have known errors where we're unable to actually establish
+		// a connection to the cluster due to say, invalid connection parameters
+		// we're going to just skip the cleanup phase since we likely won't be
+		// able to clean ourselves up anyway.
+		if internalclient.IsTerminalClientError(err) ||
+			internalclient.IsConfigurationError(err) ||
+			internalclient.IsInvalidClusterError(err) {
+			return nil
+		}
+		return err
+	}
+
+	usersClient, syncer, hasUser, err := r.userAndACLClients(ctx, user)
+	if err != nil {
+		return ignoreAllConnectionErrors(err)
+	}
+
+	if hasUser && hasManagedUser {
+		if err := usersClient.Delete(ctx, user); err != nil {
+			return ignoreAllConnectionErrors(err)
+		}
+	}
+
+	if hasManagedACLs {
+		if err := syncer.DeleteAll(ctx, user); err != nil {
+			return ignoreAllConnectionErrors(err)
+		}
+	}
+
+	return nil
+}
+
+func (r *UserReconciler) userAndACLClients(ctx context.Context, user *redpandav1alpha2.User) (*users.Client, *acls.Syncer, bool, error) {
+	usersClient, err := r.Users(ctx, user, r.extraOptions...)
+	if err != nil {
+		return nil, nil, false, err
+	}
+
+	syncer, err := r.ACLs(ctx, user, r.extraOptions...)
+	if err != nil {
+		return nil, nil, false, err
+	}
+
+	hasUser, err := usersClient.Has(ctx, user)
+	if err != nil {
+		return nil, nil, false, err
+	}
+
+	return usersClient, syncer, hasUser, nil
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *UserReconciler) SetupWithManager(ctx context.Context, mgr ctrl.Manager) error {
+	if err := registerUserClusterIndex(ctx, mgr); err != nil {
+		return err
+	}
+
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&redpandav1alpha2.User{}).
+		Owns(&corev1.Secret{}).
+		Watches(&redpandav1alpha2.Redpanda{}, handler.EnqueueRequestsFromMapFunc(func(ctx context.Context, o client.Object) []reconcile.Request {
+			requests, err := usersForCluster(ctx, r.KubernetesClient(), client.ObjectKeyFromObject(o))
+			if err != nil {
+				mgr.GetLogger().V(1).Info("skipping reconciliation due to fetching error", "error", err)
```

Comment:
An error here wouldn't actually lead to a reconciliation skippage, right? It's just indicative that the user reconciler would skip a redpanda update. Could we add a bit more context to this message to avoid any potential misleads in a debugging situation?

```suggestion
				mgr.GetLogger().V(1).Info("possibly skipping user reconciliation due to failure to fetch users associated with cluster", "error", err, "cluster", o.GetName())
```

Reply:
Yes, it'd _potentially_ skip a reconciliation run triggered by a cluster update because of an error. So, not a reconciliation run because of a changed object, but because of a changed upstream dependency.